### PR TITLE
fix(e2e): replace downtime detection with recovery-based assertions (WU-3)

### PR DESCRIPTION
**Pull request by MacAttak**
## Summary

- Replace flaky "downtime detection" pod assertions with deterministic UID-change recovery pattern
- Extract `_assert_pod_recovery()` helper used by both MinIO and Polaris tests
- Remove dead `_check_pod_terminated` function
- Add 26 unit tests covering all acceptance criteria and boundary conditions

## Acceptance Criteria

| Criterion | Status | Evidence |
|-----------|--------|----------|
| AC-1: Pod replacement detected via UID change | :white_check_mark: PASS | `_assert_pod_recovery` records UID before delete, asserts `new_uid != original_uid`. 11 tests. |
| AC-2: Service recovery within 30 seconds | :white_check_mark: PASS | `time.monotonic()` timing, default `timeout=30.0`, `logger.info` with duration. 3 tests. |
| AC-3: No downtime detection assertion | :white_check_mark: PASS | No `service_down` variable in helper. Source inspection tests verify. 2 tests. |
| AC-4: Both MinIO and Polaris tests updated | :white_check_mark: PASS | Both test methods call `_assert_pod_recovery`. Parameterized tests. 2 tests. |
| BC-1: Pod deletion fails → clear error | :white_check_mark: PASS | `pytest.fail` on None/empty UID; `AssertionError` on non-zero kubectl exit. 4 tests. |
| BC-2: Timeout → fail with message | :white_check_mark: PASS | `pytest.fail` with service name when `wait_for_condition` returns False. 2 tests. |
| BC-3: Multiple pods → handle gracefully | :white_check_mark: PASS | Uses `items[0]` jsonpath (single-replica assumption documented). 2 tests. |

## Blast Radius

- **Changed**: `tests/e2e/test_service_failure_resilience_e2e.py` (E2E resilience tests)
- **Added**: `tests/e2e/test_resilience_recovery.py` (unit tests for helper)
- **NOT changed**: Production code, other E2E tests, Helm charts, CI config
- **Scope**: Local — only affects service resilience test assertions

## Gate Results

| Gate | Status | Findings (B/W/I) |
|------|--------|-------------------|
| Build | PASS | 0/0/0 |
| Tests | PASS | 0/0/2 |
| Security | PASS | 0/0/0 |
| Wiring | PASS | 0/0/1 |
| Spec | PASS | 0/0/0 |

## Evidence

- `gate-build.md`: ruff + mypy clean
- `gate-tests.md`: 26/26 pass, all 7 previous findings resolved, mutation resistance PASS
- `gate-security.md`: No secrets, no injection, no CWE-532
- `gate-wiring.md`: No dead code, no unused imports, no architecture violations
- `gate-spec.md`: All 4 ACs and 3 BCs mapped to implementation + tests

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: tests/e2e/test_resilience_recovery.py [hunk 1]
```diff
@@ -0,0 +1,874 @@
...
+                "tests.e2e.test_service_failure_resilience_e2e.run_kubectl",
+                return_value=subprocess.CompletedProcess(
+                    args=[], returncode=0, stdout="pod deleted", stderr=""
+                ),
+            ) as mock_kubectl,
+            patch(
+                "tests.e2e.test_service_failure_resilience_e2e.wait_for_condition",
+                return_value=True,
+            ) as mock_wait,
+        ):
```
[the remark this answers is not in the document]
Comment (MacAttak) [reply tests/e2e/test_resilience_recovery.py]: Resolved in PR #191. A `recovery_mocks` fixture now patches all 3 dependencies as a dict. Individual tests override `side_effect`/`return_value` on the specific mock they need. The 15x inline 3-patch blocks are eliminated.

File: tests/e2e/test_service_failure_resilience_e2e.py [hunk 2]
```diff
@@ -373,47 +242,107 @@ def _get_pod_uid(label_selector: str) -> str | None:
...
+        response = httpx.get(url, timeout=3.0)
+        return response.status_code == 200
+    except (httpx.HTTPError, OSError):
+        return False
+
+
+logger = logging.getLogger(__name__)
+
+
+def _assert_pod_recovery(
```
[the remark this answers is not in the document]
Comment (MacAttak) [reply tests/e2e/test_service_failure_resilience_e2e.py]: Resolved in PR #191. `logger = logging.getLogger(__name__)` is now defined at line 38, immediately after imports and before any class/function definitions.

File: tests/e2e/test_resilience_recovery.py [hunk 3]
```diff
@@ -0,0 +1,874 @@
...
+@pytest.fixture(autouse=True)
+def _patch_namespace(monkeypatch: pytest.MonkeyPatch) -> None:
+    """Pin NAMESPACE so kubectl mocks match deterministically."""
+    monkeypatch.setenv("FLOE_E2E_NAMESPACE", "floe-test")
+
+
+def _import_helper() -> Any:
+    """Import _assert_pod_recovery from the E2E module.
+
+    Returns:
```
[the remark this answers is not in the document]
Comment (MacAttak) [reply tests/e2e/test_resilience_recovery.py]: Resolved in PR #191. The root cause was that `namespace` was a module-level constant evaluated at import time, making `monkeypatch.setenv` a no-op. The fix makes `namespace` an explicit parameter on all extracted functions (`get_pod_uid`, `check_pod_ready`, `assert_pod_recovery`). The `_patch_namespace` fixture and `_import_helper()` function have been removed entirely — tests pass the namespace directly.

**Comment by MacAttak** [comment]
**Responding to Greptile's out-of-diff findings:**

**1. `recovery_seconds` timer starts before deletion (P2)**

Addressed in PR #191. The timer placement is intentional in the extracted `assert_pod_recovery` — it measures total wall-clock time including the delete RPC, which is what callers care about for operational budgeting. However, the E2E test assertions (`assert recovery_secs < 30.0`) now compare against the `assert_pod_recovery` return value which includes this overhead, and the 30s limit in the E2E tests has always been generous enough to absorb 1-2s of kubectl latency. The extracted function uses `time.monotonic()` correctly.

**2. Redundant `_get_pod_uid` call after `wait_for_condition` (P2)**

Acknowledged — the extra kubectl round-trip and TOCTOU window are valid observations. In PR #191, the extracted `assert_pod_recovery` in `testing/fixtures/kubernetes.py` retains this pattern for simplicity (the closure approach adds complexity for a theoretical edge case in test infrastructure). The extra call is a single `kubectl get` that takes <100ms. If this becomes a practical issue, the UID-caching approach can be added as a follow-up.
